### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by AlselwiAisha**
- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
- Do not use the bullet gem in this exercise - instead analyze the logs as showed in this [article](https://dev.to/junko911/rails-n-1-queries-and-eager-loading-10eh)
- In your PR description include the screenshots of your app's logs before and after the fix
    **screenshots of our app's logs before the fix:**
Befor

    **screenshots of our app's logs after the fix:**
After

 **Integration specs**
- Use Capybara to write integration tests for each view in your project.
- Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.
- User index page:
    - I can see the username of all other users.
    - I can see the profile picture for each user.
    - I can see the number of posts each user has written.
    - When I click on a user, I am redirected to that user's show page.
- user show page:
    - I can see the user's profile picture.
    - I can see the user's username.
    - I can see the number of posts the user has written.
    - I can see the user's bio.
    - I can see the user's first 3 posts.
    - I can see a button that lets me view all of a user's posts.
    - When I click a user's post, it redirects me to that post's show page.
    - When I click to see all posts, it redirects me to the user's post's index page.
- User post index page:
    - I can see the user's profile picture.
    - I can see the user's username.
    - I can see the number of posts the user has written.
    - I can see a post's title.
    - I can see some of the post's body.
    - I can see the first comments on a post.
    - I can see how many comments a post has.
    - I can see how many likes a post has.
    - I can see a section for pagination if there are more posts than fit on the view.
    - When I click on a post, it redirects me to that post's show page.
- Post show page:
    - I can see the post's title.
    - I can see who wrote the post.
    - I can see how many comments it has.
    - I can see how many likes it has.
    - I can see the post body.
    - I can see the username of each commentor.
    - I can see the comment each commentor left.

### CODE REVIEW

**Review by skyv26 — COMMENTED** [review]
## Hi @HenaryKhuraijam, 

While you made a great effort in this project, I cannot proceed to review your code.

## Invalid Code Review Request 🟨 



## Reason 🗒️ 
You have submitted the wrong project PR, as per the requirement you should open the PR in your repository that contains the commits from your partner, this time the milestone is a little different, please read the `requirements` carefully and the most important thing, which is another reason of making this PR invalidation is your contribution is significantly very less. 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

#### Cheers, and Happy coding!👏👏👏

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: spec/view/users/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    end
+
+    it "should have a button to view all of a user's posts" do
+      expect(page).to have_link('', href: user_posts_path(@user))
+    end
+
+    it 'should redirects to all posts show page.' do
+      click_link(@first_post.title)
+      expect(page).to have_current_path(user_post_path(@user, @first_post))
+    end
```
Comment (omar-muhamad) [inline spec/view/users/user_show_spec.rb]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20to%20see%20all%20posts%2C%20it%20redirects%20me%20to%20the%20user%27s%20post%27s%20index%20page.), Please add a test that redirects to the user's post's index page when you click on a `see all posts`.

File: spec/view/users/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    it "should render user's first 3 posts" do
+      expect(page).to have_content(@first_post.text)
+      expect(page).to have_content(@second_post.text)
+      expect(page).to have_content(@third_post.text)
+    end
+
+    it "should have a button to view all of a user's posts" do
+      expect(page).to have_link('', href: user_posts_path(@user))
+    end
```
Comment (omar-muhamad) [inline spec/view/users/user_show_spec.rb]: - [ ] For this test, you should be more specific about the matcher option you added to this test. Please add the link text `see all posts` to the matcher.
```suggestion
    it "should have a button to view all of a user's posts" do
      expect(page).to have_link('see all posts', href: user_posts_path(@user))
    end
```

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi Team,

### Nice job, you are almost there.

cat-driving-serious

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/view/users/user_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    it "should have a button to view all of a user's posts" do
+      expect(page).to have_link('see all posts', href: user_posts_path(@user))
+    end
+
+    it 'should redirects to all posts show page.' do
+      click_link(@first_post.title)
+      expect(page).to have_current_path(user_post_path(@user, @first_post))
+    end
```
Comment (thecodechaser) [inline spec/view/users/user_show_spec.rb:46]: - [ ] As requested by the previous reviewer, Please make sure to add the below test to the user show page and it should pass without any errors.
   - Tests that redirects to the user's post's index page when you click on a see all posts.

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @AlselwiAisha, and Team👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight 🎉
- Good job fixing n+1 problem. ✔️ 
- Good job with integration tests. ✔️
- No linter errors. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- PR has a good title and summary. ✔️



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @AlselwiAisha, and Team👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

image

## To Highlight 🎉
- All tests are passed. ✔️
- Good job fixing n+1 problem. ✔️
- No linter errors. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**